include/RAJA/policy/cuda/multi_reduce.hpp

```diff
@@ -0,0 +1,764 @@
+    tally_rep = omp_get_thread_num();
+#endif
+    int tally_offset = GetTallyOffset{}(bin, m_tally_bins, tally_rep, m_tally_replication);
+    Combiner{}(m_tally_mem[tally_offset], value);
+  }
+
+private:
+  using shared_tuning = typename tuning::SharedAtomicReplicationTuning;
+  using SharedAtomicReplicationConcretizer = typename shared_tuning::AtomicReplicationConcretizer;
+  using GetSharedIndex = typename shared_tuning::ReplicationIndexer;
+  using GetSharedOffset_rebind = typename shared_tuning::OffsetCalculator;
+  using GetSharedOffset = typename GetSharedOffset_rebind::template rebind<int>;
+
+  using typename TallyData::GetTallyIndex;
+  using typename TallyData::GetTallyOffset;
+
+
+  static constexpr int s_shared_offset_unknown = std::numeric_limits<int>::max();
+  static constexpr int s_shared_offset_invalid = std::numeric_limits<int>::max() - 1;
+
+
+  using TallyData::m_tally_mem;
+  using TallyData::m_identity;
+  using TallyData::m_num_bins;
+  using TallyData::m_tally_bins;
+  using TallyData::m_tally_replication;
+
+  int m_shared_offset; // in bytes
+  int m_shared_replication; // power of 2
+
+
+  RAJA_DEVICE
+  T* get_shared_mem() const
+  {
+    if (m_shared_offset == s_shared_offset_invalid) {
+      return nullptr;
+    }
+    extern __shared__ char shared_mem[];
+    return reinterpret_cast<T*>(&shared_mem[m_shared_offset]);
+  }
+};
+
+
+/*!
+ **************************************************************************
+ *
+ * \brief  Cuda multi-reduce data class template.
+ *
+ * This class manages synchronization, data lifetimes, and interaction with
+ * the runtime kernel launch info passing facilities.
+ *
+ * This class manages the lifetime of underlying reduce_data_type using
+ * calls to setup and teardown methods. This includes storage durations:
+ * - permanent, the lifetime of the parent object
+ * - launch, setup before a launch using the launch parameters and
+ *           teardown after the launch
+ * - device, setup all device threads in a kernel before any block work and
+ *           teardown all device threads after all block work is finished
+ *
+ **************************************************************************
+ */
+template < typename T, typename t_MultiReduceOp, typename tuning >
+struct MultiReduceDataCuda
+{
+  static constexpr bool atomic_available = RAJA::reduce::cuda::cuda_atomic_available<T>::value;
+
+  //! cuda reduction data storage class and folding algorithm
+  using reduce_data_type =
+      std::conditional_t<(atomic_available),
+        std::conditional_t<(tuning::algorithm == multi_reduce_algorithm::init_host_combine_block_atomic_then_grid_atomic),
+          cuda::MultiReduceBlockThenGridAtomicHostInit_Data<t_MultiReduceOp, T, tuning>,
+          std::conditional_t<(tuning::algorithm == multi_reduce_algorithm::init_host_combine_global_atomic),
+            cuda::MultiReduceGridAtomicHostInit_Data<t_MultiReduceOp, T, tuning>,
+            void>>,
+      void>;
```

Comment:
This seems pretty key for connecting this lifetime management object to one of the data storage objects with `setup_device` capability above. How does this bridge the two classes? What's an example of a type that gets produced from this fold?

Reply:
This just chooses between `cuda::MultiReduceBlockThenGridAtomicHostInit_Data<t_MultiReduceOp, T, tuning>` and `cuda::MultiReduceGridAtomicHostInit_Data<t_MultiReduceOp, T, tuning>` based on `tuning::algorithm`.

Reply:
Ah, that makes sense. That's why this lifetime class has its copy constructors deleted, while the data classes have their copy constructors default'ed.